feat: new 'IKangle' converter , inverse kinematics bone

zaynopenapp opened the pull request:
The new 'IK angle' converter is a simple inverse kinematics solution. In line with the issues discussed in https://github.com/synfig/synfig/issues/3571 or https://forums.synfig.org/t/inverse-kinematic-for-bones/3068, and (https://forums.synfig.org/t/ik-system/4132)it is hoped that inverse kinematics will finally be available for users who have long wanted Synfig to have IK bones.

forum: https://forums.synfig.org/t/bone-like-movement/588, https://forums.synfig.org/t/skeleton-bones-duplication/3650, https://forums.synfig.org/t/inverse-kinematic-for-bones/3068


sample here. https://forums.synfig.org/t/ik-angle-converter-test/16552 and https://youtu.be/kgnKCwXtjIQ

The basic idea behind my **IK Angle** converter is based on the geometric intersection of two circles.

Imagine a simple two-bone chain. The first circle is centered at the root bone with a radius equal to the length of the first bone. The second circle is centered at the target with a radius equal to the length of the second bone.

The intersection of these two circles gives the possible position of the intermediate joint (for example, the elbow or the knee). Once that joint position is known, the rotation angle of each bone can be calculated directly using vector angles.

Because this approach is based on geometry rather than an iterative solver, it produces a stable solution with very little computation. There is no need for repeated iterations such as those used by CCD or FABRIK for a simple two-bone chain.

To support different bending directions, the solver simply chooses one of the two possible intersection points (for example, left/right or up/down), depending on the desired pole direction.

This geometric approach is the core idea behind my IK Angle converter.

The distance between the root and the target determines whether the circles intersect. When they do, there are generally two possible solutions, corresponding to the two possible elbow or knee positions. The converter selects one of these solutions and computes the bone rotations from the resulting vectors.

zaynopenapp committed:
Update POTFILES.in

ik angle converter

zaynopenapp committed:
Update CMakeLists.txt

ik angle converter

zaynopenapp committed:
Update Makefile_insert

Inverse Kinematic converter

zaynopenapp committed:
ik angle file .cpp and .h

IK angle converter. new converter for native IK 2025

zaynopenapp committed:
Update valuenode_ik_angle.cpp

zaynopenapp committed:
Update valuenode_ik_angle.h

fix enum RigType , enum TargetBone

zaynopenapp reviewed (commented):
Why my code error?

rodolforg commented:
> Why my code error?

<img width="1187" height="705" alt="image" src="https://github.com/user-attachments/assets/5bb43b66-fdc0-4188-855a-7d0f67cd4baf" />

zaynopenapp commented:
Ok. Thank you. I will immediately fix the part that has the error.

_**Edit BobSynfig: The image was not displayed in the quote, still taking empty space**_

zaynopenapp committed:
Update valuenode_ik_angle.cpp

fix type

zaynopenapp reviewed (commented):
I’ve fixed it according to Gemini’s suggestions, though not all of them. I’ve also tested the Windows artifact, and it runs just like the Linux version I built myself.
<img width="1920" height="1080" alt="ik angle" src="https://github.com/user-attachments/assets/9775f382-cd1e-4314-95af-2bdd9654e4f9" />

zaynopenapp committed:
color bone script add in header

Added a new parameter 'color_bones' to Layer_Skeleton.

zaynopenapp committed:
Add color parameter handling in Layer_Skeleton

zaynopenapp commented:
Below is a comparison of files from an IK example using a plugin and an IK angle converter:

<img width="681" height="254" alt="ik file compare" src="https://github.com/user-attachments/assets/8bfc8011-1832-46ef-8291-cd49816a2154" />

Although compressing into a .sifz file reduces the file size, the original file still contains a large number of converters, which significantly impacts Synfig’s performance. This is only for a single IK rig object (a dinosaur). One can imagine how heavy it would become with 5 such objects using the IK plugin.

[comparefile.zip](https://github.com/user-attachments/files/26275687/comparefile.zip)

zaynopenapp committed:
Update valuenode_ik_angle.cpp

There is an update to the IK angle converter. It can now determine the positions of elbow 1 and elbow 2 (3 joint bones).
Of course, it requires another converter such as a vector angle. It can also stretch the bones by calculating data from the bone length

SlamRockwell commented:
This would be amazing to have in Synfig and I hope the developers can pull it in soon! :)

BobSynfig commented:
AppImage for test [UNOFFICIAL-1.5.5-2026.05.06-test-pr-3623](https://github.com/BobSynfig/synfig/releases/tag/UNOFFICIAL-1.5.5-2026.05.06-test-pr-3623)

zaynopenapp commented:
You can find the documentation for the **IK angle** converter here. Thank you.
[link](https://github.com/zaynopenapp/IK-angle-new-converter-Synfig)

<img width="1961" height="1321" alt="ikangle_tuto" src="https://github.com/user-attachments/assets/1ccda642-2db6-4ff2-8487-71708cc8bade" />

zaynopenapp commented:
> This would be amazing to have in Synfig and I hope the developers can pull it in soon! :)

[@SlamRockwell](https://github.com/SlamRockwell) Thank you for your support. you can review this PR

zaynopenapp committed:
Refactor IK angle calculations for clarity and efficiency

Refactor: simplify the implementation without changing behavior.

Svarov-RZM commented:
I see I've been tagged on this, but it's a bit too much for me to handle (I am a sysadmin, remember).
Also, I don't use bones in Synfig so I won't be able to do a good test.

I can do a surface level test of this PR if you want me to, but the code review **must** be done by someone more competent.

zaynopenapp reviewed (commented):
I fixed this issue
